ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala

```diff
@@ -196,45 +196,49 @@ class InMemoryLedger(
-              party -> Filters(if (templates.nonEmpty) Some(InclusiveFilters(templates)) else None)
-          }))
-          .toList))
-      .map { contract =>
-        GetActiveContractsResponse(
-          workflowId = contract.workflowId.getOrElse(""),
-          activeContracts = List(
-            CreatedEvent(
-              contract.eventId,
-              contract.id.coid,
-              Some(LfEngineToApi.toApiIdentifier(contract.contract.template)),
-              contractKey = contract.key.map(
-                ck =>
+  ): (Source[GetActiveContractsResponse, NotUsed], Offset) =
+    (
+      Source
+        .fromIterator[ActiveContract](
+          () =>
+            acs.activeContracts.valuesIterator.flatMap(
+              index
+                .EventFilter(_)(TransactionFilter(filter.map {
+                  case (party, templates) =>
+                    party -> Filters(
+                      if (templates.nonEmpty) Some(InclusiveFilters(templates)) else None)
+                }))
+                .toList))
+        .map { contract =>
+          GetActiveContractsResponse(
+            workflowId = contract.workflowId.getOrElse(""),
+            activeContracts = List(
+              CreatedEvent(
+                contract.eventId,
+                contract.id.coid,
+                Some(LfEngineToApi.toApiIdentifier(contract.contract.template)),
+                contractKey = contract.key.map(
+                  ck =>
+                    LfEngineToApi.assertOrRuntimeEx(
+                      "converting stored contract",
+                      LfEngineToApi
+                        .lfContractKeyToApiValue(verbose = verbose, ck))),
+                createArguments = Some(
                   LfEngineToApi.assertOrRuntimeEx(
                     "converting stored contract",
                     LfEngineToApi
-                      .lfContractKeyToApiValue(verbose = verbose, ck))),
-              createArguments = Some(
-                LfEngineToApi.assertOrRuntimeEx(
-                  "converting stored contract",
-                  LfEngineToApi
-                    .lfValueToApiRecord(verbose = verbose, contract.contract.arg.value))),
-              contract.signatories.union(contract.observers).intersect(filter.keySet).toSeq,
-              signatories = contract.signatories.toSeq,
-              observers = contract.observers.toSeq,
-              agreementText = Some(contract.agreementText)
+                      .lfValueToApiRecord(verbose = verbose, contract.contract.arg.value))),
+                contract.signatories.union(contract.observers).intersect(filter.keySet).toSeq,
+                signatories = contract.signatories.toSeq,
+                observers = contract.observers.toSeq,
+                agreementText = Some(contract.agreementText)
+              )
             )
           )
-        )
-      }
+        },
+      ledgerEnd /*TODO SC derive from copy of acs var instead*/ )
```

Comment:
What you can do is something like this at the beginning:
```
val (activeContracts, ledgerEnd) = this.synchronized { (acs.activeContracts, ledgerEnd) }
```

This way `ledgerEnd` and `acs` will be "in sync".

Reply:
With that change, the test no longer finds a duplicate.

Reply:
907a64122e